### PR TITLE
[CDAP-12293][CDAP-12301] Add app info and schedule status in ScheduleDetail

### PR DESCRIPTION
**Pull request by ecfm**
Also add REST API to filter schedules by status

https://issues.cask.co/browse/CDAP-12293
https://issues.cask.co/browse/CDAP-12301

https://builds.cask.co/browse/CDAP-RUT1239-6

### CODE REVIEW

File: cdap-proto/src/test/java/co/cask/cdap/proto/ProtoTriggerCodecTest.java [hunk 1]
```diff
@@ -67,16 +68,16 @@ private void testTriggerCodec(ProtoTrigger trigger) {
...
+                                               new ScheduleProgramInfo(SchedulableProgramType.WORKFLOW, "ww"),
+                                               ImmutableMap.of("prop3", "abc"),
+                                               new ProtoTrigger.PartitionTrigger(new DatasetId("test1", "pdfs1"), 1),
+                                               ImmutableList.<Constraint>of(), null, "DISABLED");
+    ScheduleDetail sched2 = new ScheduleDetail("default", "app1", "1.0.0", "schedone", "one time schedule",
                                                  new ScheduleProgramInfo(SchedulableProgramType.WORKFLOW, "wf112"),
                                                  ImmutableMap.of("prop", "all"),
                                                  new ProtoTrigger.TimeTrigger("* * * 1 1"),
-                                                 ImmutableList.<Constraint>of(), null);
+                                                 ImmutableList.<Constraint>of(), null, "DISABLED");
```
Comment (sameetandpotatoes) [inline cdap-proto/src/test/java/co/cask/cdap/proto/ProtoTriggerCodecTest.java]: Why "DISABLED"? Isn't it SUSPENDED or SCHEDULED?

File: cdap-proto/src/main/java/co/cask/cdap/proto/ProtoTrigger.java [hunk 2]
```diff
@@ -334,6 +334,7 @@ public String toString() {
     protected final ProgramId programId;
     protected final Set<ProgramStatus> programStatuses;
 
+
```
Comment (sameetandpotatoes) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ProtoTrigger.java]: unneeded

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java [hunk 3]
```diff
@@ -499,6 +519,24 @@ public ProgramScheduleRecord getScheduleRecord(ScheduleId scheduleId) throws Not
     return result;
   }
 
+  private List<ProgramScheduleRecord> listScheduleRecords(ApplicationId appId, @Nullable ProgramId programId) {
```
Comment (sameetandpotatoes) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java]: I would add a Javadoc for this method, just so you can explain why `programId` can be null.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 4]
```diff
@@ -790,6 +799,23 @@ protected void doGetSchedules(HttpResponder responder, String namespace, String
     }
   }
 
+  private Collection<ProgramScheduleRecord> filterSchedulesByStatus(Collection<ProgramScheduleRecord> scheduleRecords,
+                                                                    String scheduleStatus) {
```
Comment (sameetandpotatoes) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Should annotate this with Nullable then since `scheduleStatus` can be null.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 5]
```diff
@@ -790,6 +799,23 @@ protected void doGetSchedules(HttpResponder responder, String namespace, String
...
+    // No need to filter by status if scheduleStatus is null, directly return the original records
+    if (scheduleStatus == null) {
+      return scheduleRecords;
+    }
+    List<ProgramScheduleRecord> filteredRecords = new ArrayList<>();
```
Comment (sameetandpotatoes) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Use `Collections2.filter` instead?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java [hunk 6]
```diff
@@ -84,11 +84,14 @@ public static String triggerKeyForPartition(DatasetId datasetId) {
     return "partition:" + datasetId.getNamespace() + '.' + datasetId.getDataset();
   }
 
-  public static Set<String> triggerKeysForProgramStatus(ProgramId programId, Set<ProgramStatus> programStatuses) {
+  public static String triggerKeyForProgramStatus(ProgramId programId, ProgramStatus programStatus) {
```
Comment (sameetandpotatoes) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java:87]: Should rebase now that this method was already added in the implement program status scheduling PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 7]
```diff
@@ -624,7 +626,8 @@ public void getProgramStatusSchedules(HttpRequest request, HttpResponder respond
                                         @QueryParam("trigger-app-version") String triggerAppVersion,
                                         @QueryParam("trigger-program-type") String triggerProgramType,
                                         @QueryParam("trigger-program-name") String triggerProgramName,
-                                        @QueryParam("trigger-program-statuses") String triggerProgramStatuses)
+                                        @QueryParam("trigger-program-statuses") String triggerProgramStatuses,
+                                        @QueryParam("schedule-status") String scheduleStatus)
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:634]: nit: update javadoc for this param.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 8]
```diff
@@ -638,7 +641,7 @@ public void getProgramStatusSchedules(HttpRequest request, HttpResponder respond
     final ProgramId triggerProgramId = appId.program(ProgramType.valueOfCategoryName(triggerProgramType),
                                                      triggerProgramName);
     programScheduler.findSchedules(triggerProgramId.toString());
-    Set<ProgramScheduleRecord> schedules = new HashSet<>();
+    Collection<ProgramScheduleRecord> schedules = new HashSet<>();
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:648]: why is this changed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java [hunk 9]
```diff
@@ -481,6 +501,10 @@ public ProgramScheduleRecord getScheduleRecord(ScheduleId scheduleId) throws Not
 
   /*------------------- private helpers ---------------------*/
 
+  /**
+   * List schedules in a given application and optionally filtered by the program triggered by the schedule
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java]: nit: `filter by`

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java]: also the comment is not super clear.. this method gets the schedules in the application and if program id is provided it will only get schedules associated with that program right? we should not say `triggered` then..right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java [hunk 10]
```diff
@@ -500,6 +524,28 @@ public ProgramScheduleRecord getScheduleRecord(ScheduleId scheduleId) throws Not
   }
 
   /**
+   * List schedule records in a given application and optionally filtered by the program triggered by the schedule
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java]: same here.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java (hunk 6 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java:87]: is this already addressed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java [hunk 11]
```diff
@@ -99,6 +99,16 @@ public ProgramScheduleStatus getScheduleStatus(ScheduleId scheduleId) throws Not
   }
 
   @Override
+  public List<ProgramScheduleRecord> listScheduleRecords(ApplicationId appId) throws NotFoundException {
+    return null;
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java]: better return empty list?

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java [hunk 12]
```diff
@@ -99,6 +99,16 @@ public ProgramScheduleStatus getScheduleStatus(ScheduleId scheduleId) throws Not
...
+  }
+
+  @Override
+  public List<ProgramScheduleRecord> listScheduleRecords(ProgramId programId) throws NotFoundException {
+    return null;
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java]: same here

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java [hunk 13]
```diff
@@ -99,6 +99,16 @@ public ProgramScheduleStatus getScheduleStatus(ScheduleId scheduleId) throws Not
...
+  }
+
+  @Override
   public Collection<ProgramScheduleRecord> findSchedules(String triggerKey) {
     return null;
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java]: here as well

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java (hunk 12 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NoOpScheduler.java]: in general better to return empty list rather than null so that client does not have to do the null checks.

**Review by sagarkapare — COMMENTED** [review]
mostly looks good. minor comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 8 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:648]: so that the result from `filterSchedulesByStatus` method can be assigned to it, which returns a Collection

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java (hunk 6 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java:87]: addressed

File: cdap-proto/src/main/java/co/cask/cdap/proto/ScheduleDetail.java [hunk 14]
```diff
@@ -40,28 +40,55 @@
...
+  private final String application;
+  private final String applicationVersion;
   private final String name;
   private final String description;
   private final ScheduleProgramInfo program;
   private final Map<String, String> properties;
   private final Trigger trigger;
   private final List<? extends Constraint> constraints;
   private final Long timeoutMillis;
+  private final String scheduleStatus;
```
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ScheduleDetail.java]: nit: better call it `status` since we are already in `ScheduleDetail` class?

**Review by sagarkapare — COMMENTED** [review]
one minor question. otherwise LGTM 👍

**Comment by ecfm** [comment]
Build is green. Squash and merge.
